test/test_cpp_api_parity.py

```diff
@@ -0,0 +1,185 @@
+    cpp_output.sizes().vec() == python_output.sizes().vec() &&
+    cpp_output.allclose(python_output),
+    parity_test_error_msg, ": forward output doesn't match");
+}
+""")
+
+        def process_test_params(test_params, python_module_class, cpp_namespace):
+            module_name = test_params['module_name']
+            desc = test_params.get('desc', None)
+            test_params['module_variant_name'] = module_name + (('_' + desc) if desc else '')
+            test_params['python_constructor_args'] = test_params.get('constructor_args')
+            test_params['cpp_constructor_args'] = test_params.get('cpp_constructor_args')
+            test_params['cpp_namespace'] = cpp_namespace
+            test_params['cpp_source'] = test_params.get('cpp_source', '')
+            test_params['expect_error'] = test_params.get('expect_error', False)
+            if not python_module_class:
+                python_module_class = getattr(torch.nn, module_name)
+            test_params['python_module_class'] = python_module_class
+            return test_params
+
+        test_params_map = {}
+        for test_params in module_tests:
+            test_params = process_test_params(test_params, python_module_class, cpp_namespace)
+            module_name = test_params['module_name']
+            if module_name not in test_params_map:
+                test_params_map[module_name] = []
+            test_params_map[module_name].append(test_params)
+
+        # Generate C++ code for each test case
+        cpp_source = TORCH_NN_MODULE_COMMON_TEST_HARNESS + test_params['cpp_source']
+        functions = []
+        for module_name in module_names:
+            for test_params in test_params_map[module_name]:
+                cpp_source += TORCH_NN_MODULE_WRAPPER.substitute(
+                    module_variant_name=test_params['module_variant_name'],
+                    module_qualified_name=test_params['cpp_namespace'] + test_params['module_name'],
+                    cpp_constructor_args=test_params['cpp_constructor_args'])
+                for method in torch_nn_test_methods:
+                    functions.append(test_params['module_variant_name'] + '_test_' + method)
+
+        # Just-in-time compile the C++ test code
+        cpp_module = torch.utils.cpp_extension.load_inline(
+            name=test_suite_name,
+            cpp_sources=cpp_source,
+            functions=functions,
+            verbose=True,
+        )
+
+        def test_method(method_name, test_params):
+            expect_error = test_params['expect_error']
+            input_size = test_params['input_size']
+            python_module_class = test_params['python_module_class']
+            python_constructor_args = test_params['python_constructor_args']
+            module_variant_name = test_params['module_variant_name']
+            test_name = module_variant_name + '_test_' + method_name
+            example_input = torch.randn(input_size)
+            with common.freeze_rng_state():
```

Comment:
This is unnecessary now that you're doing the manual seed.

You could also use `with common.freeze_rng_state():` to make the code cleaner; something like the following might work:
```
with common.freeze_rng_state():
    python_output = ...
with common.freeze_rng_state():
    check_cpp_output...
```